design-proposals/klip-1-improve-udf-interfaces.md

```diff
@@ -0,0 +1,291 @@
+    agg.put("count", agg.get("count") + 1);
+    return agg;
+  }
+  
+  @Override
+  public Struct merge(final Struct agg1, final Struct agg2) {
+    agg1.put("sum", agg1.get("sum") + agg2.get("sum"));
+    agg1.put("count", agg1.get("count") + agg2.get("count"));
+    return agg1;
+  }
+ 
+  /**
+  * This is the new method that overrides {@code Exportable#export(A agg)} (note that
+  * {@code Exportable<A,VR>} is defined with type parameters for the input and output
+  * of this method.
+  */
+  @Override
+  public Integer export(final Struct agg) {
+    return agg.getInt64("sum") / agg.getInt64("count");
+  }
+  
+}
+```
+
+**NOTE:** It should be possible to support arbitrary Java objects for aggregation without much
+extra work. For the scope of this KLIP, however, supporting just KSQL types (including `Struct`) 
+should suffice.
+
+#### Alternative
+
+If we expect UDAFs to commonly require exportable functionality, then we can make this change in a
+backwards incompatible change and introduce the `export` (or `terminate`) method into the UDAF
+interface directly.
+
+## Test plan
+
+Each of these improvements will come with some built-in UDF/UDAF examples. These examples will then
+be comprehensively covered by both unit tests as well as corresponding Query Translation Tests.
+Specifically, the UDFs described in the [Value](#value) section will all be implemented alongside
+this KLIP. For the new UDAF complex aggregation, we will ensure that the generated topologies do not
+include any unexpected steps such as a repartition.
+
+## Documentation Updates
+
+* The `Example UDF Class` in `udf.rst` will be extended to showcase the new features. Namely, we
+will replace the multiply double method to include varargs:
+
+>```java
+>      @Udf(description = "multiply N non-nullable DOUBLEs.")
+>      public double multiply(final double... values) {
+>        return Arrays.stream(values).reduce((a, b) -> a * b);;
+>      }
+>```
+
+* The `Supported Types` section in `udf.rst` will include `Struct` and have an updated note at the 
+bottom which reads:
+
+> Note: Complex types other than List, Map and Struct are not currently supported. Generic types
+> for List and Map are supported, but component elements must be of one of the supported types. If
+> the types are generic, the output type must be able to be inferred from one or more of the input
+> parameters.
+
+* The `UDAF` section in `udf.rst` will have an additional section to describe usage of the 
+`Exportable` interface:
+
+> A UDAF can support complex aggregation types if it implements `io.confluent.common.Exportable`.
+> This will allow your custom UDAF to convert some intermediate type used for aggregation to a 
+> KSQL output value. For example:
+>
+>```java
+>@UdfFactory(description = "Computes a running average")
+>class AverageUdaf implements Udaf<Long, Struct>, Exportable<Struct, Long> {
+>   
+>  @Override
+>  public Long export(Struct runningSum) {
```

Comment:
nit, also applicable to the proposed code implementation of 'average' above: i found it harder to follow the example because i kept wondering where you were going to allow for fractional numbers in the output :)  - if we change the return type to Double then it will help clarify which `Long` in the various signatures represents the incoming values to aggregate vs which one is the return type.